File: common/src/main/java/com/ssblur/scriptor/feature/ScriptorFeatures.java
package com.ssblur.scriptor.feature;

import com.ssblur.scriptor.ScriptorMod;
import dev.architectury.registry.level.biome.BiomeModifications;
import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.BiomeTags;
import net.minecraft.world.level.levelgen.GenerationStep;
import net.minecraft.world.level.levelgen.feature.Feature;
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;

public class ScriptorFeatures {
  public static final DeferredRegister<Feature<?>> FEATURES = DeferredRegister.create(ScriptorMod.MOD_ID, Registries.FEATURE);

  public static final RegistrySupplier<Feature<?>> ENGRAVING_FEATURE = FEATURES.register("engraving",
    () -> new ScriptorInscriptionsFeature(NoneFeatureConfiguration.CODEC)
  );

  public static void register() {
    FEATURES.register();
    BiomeModifications.addProperties(
      biome -> biome.hasTag(BiomeTags.IS_OVERWORLD),
      (biomeContext, mutable) ->
        mutable.getGenerationProperties().addFeature(
          GenerationStep.Decoration.RAW_GENERATION,
          ResourceKey.create(Registries.PLACED_FEATURE, ResourceLocation.tryBuild("scriptor", "engravings"))
    ));
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/feature/ScriptorFeatures.kt
package com.ssblur.scriptor.feature

import com.ssblur.scriptor.ScriptorMod
import com.ssblur.scriptor.block.GenerateBlock
import com.ssblur.unfocused.extension.BlockStateExtension.matches
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.levelgen.feature.Feature
import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration

@Suppress("unused")
object ScriptorFeatures {
  val ENGRAVING_FEATURE = ScriptorMod.registerFeature("engraving") {
    object: Feature<NoneFeatureConfiguration?>(NoneFeatureConfiguration.CODEC) {
      override fun place(context: FeaturePlaceContext<NoneFeatureConfiguration?>): Boolean {
        if (context.level().getBlockState(context.origin().offset(0, -1, 0)) matches Blocks.WATER)
          return false
        return context.level().setBlock(context.origin(), GenerateBlock.generateEngraving(), 0)
      }
    }
  }

  fun register() {}
}
